src/Draco.Compiler.Internal/Utilities/Result.cs

```diff
@@ -0,0 +1,157 @@
+using System;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Draco.Compiler.Internal.Utilities;
+
+/// <summary>
+/// A union type containing either the result of an operation or an error.
+/// </summary>
+/// <typeparam name="T">The type of the inner value.</typeparam>
+/// <typeparam name="TError">The type of the inner error.</typeparam>
+public readonly struct Result<T, TError>
+{
+    private readonly T value;
+    private readonly TError error;
+    private readonly bool isSuccess;
+
+    /// <summary>
+    /// The value of the result. Throws an exception if the result is not a value.
+    /// </summary>
+    /// <exception cref="InvalidOperationException"></exception>
+    public T? Value =>
+        this.isSuccess
```

Comment:
We usually keep the condition in the same line for ternary, less overall indentation